pkg/server/documents_operations.go

```diff
@@ -165,3 +164,79 @@ func (s *ImmuServer) CollectionUpdate(ctx context.Context, req *documentschema.C
 	}
 	return resp, nil
 }
+
+func (s *ImmuServer) DocumentSearch(ctx context.Context, req *documentschema.DocumentSearchRequest) (*documentschema.DocumentSearchResponse, error) {
+	db, err := s.getDBFromCtx(ctx, "DocumentSearch")
+	if err != nil {
+		return nil, err
+	}
+
+	// get the session from the context
+	sessionID, err := sessions.GetSessionIDFromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+	sess, err := s.SessManager.GetSession(sessionID)
+	if err != nil {
+		return nil, err
+	}
+
+	// generate a unique query name based on the request parameters
+	queryName := generateQueryName(req)
+
+	// check if the paginated reader for this query has already been created
+	var resultReader sql.RowReader
+	var pgreader *sessions.PaginatedReader
+
+	pgreader, err = sess.GetPaginatedReader(queryName)
+	if err != nil { // paginated reader does not exist, create a new one and add it to the session
+		resultReader, err = db.SearchDocuments(ctx, req)
+		if err != nil {
+			return nil, err
+		}
+
+		// store the reader in the session for future use
+		pgreader = &sessions.PaginatedReader{
+			Reader:         resultReader,
+			LastPageNumber: req.Page,
+			LastPageSize:   req.PerPage,
+		}
+		sess.SetPaginatedReader(queryName, pgreader)
+	} else { // paginated reader already exists, resume reading from the correct offset based on pagination parameters
+		// do validation on the pagination parameters
+		if req.Page < pgreader.LastPageNumber {
+			return nil, ErrInvalidPreviousPage
+		}
+		resultReader = pgreader.Reader
+	}
+
+	// read the next page of data from the paginated reader
+	results, err := document.ReadStructMessagesFromReader(ctx, resultReader, int(req.PerPage))
+	if err != nil && err != sql.ErrNoMoreRows {
+		return nil, err
+	}
+
+	// update the pagination parameters for this query in the session
+	sess.UpdatePaginatedReader(queryName, req.Page, req.PerPage, int(pgreader.TotalRead)+len(results))
+
+	if err == sql.ErrNoMoreRows {
+		// end of data reached, remove the paginated reader and pagination parameters from the session
+		delErr := sess.DeletePaginatedReader(queryName)
+		s.Logger.Errorf("error deleting paginated reader: %v", delErr)
+	}
+
+	return &documentschema.DocumentSearchResponse{
+		Results: results,
+	}, err
+}
+
+func generateQueryName(req *documentschema.DocumentSearchRequest) string {
```

Comment:
the idea of generating an identifier from the request is interesting but it might lead to unexpected results e.g. same query requested in the session at different ranges (offsets), that query will be considered to be the same by the server.
Probably it's safer to have a unique id for the reader that it has to be provided in the next requests

Reply:
makes sense, I was assuming the requests would be unique, but this method sounds better, added it